src/models/ConstrainedMetaModel.ts

```diff
@@ -0,0 +1,93 @@
+import { MetaModel } from './MetaModel';
+
+export class ConstrainedMetaModel extends MetaModel {
+  constructor(
+    public name: string,
+    public originalInput: any, 
+    public type: string) {
+    super(name, originalInput);
+  }
+}
+
+export class ConstrainedReferenceModel extends ConstrainedMetaModel {
+  constructor(
+    public name: string,
+    public originalInput: any, 
+    public type: string, 
```

Comment:
As I wrote in another comment, you don't need to redefine properties:

```suggestion
    name: string,
    originalInput: any, 
    type: string, 
```